### ISSUE TITLE
Signing a UWP Application using either VSBuild or MSBuild not working as expected

### ISSUE THREAD

**Issue by luEbe** [actions/virtual-environments#3183]
**Description**  

We used the VSBuild task to build a UWP application and sign it with a certificate that was uploaded to the _Secure Files Storage_ on Azure and then loaded into the certificate store using a Script utilizing the PowerShellV2 Task.
The build is successful but the application is not signed with that certificate, rather it seems like, it is just using the information given in the appxmanifest file to create a self-signed certificate.

After some attempts we replaced the **VSBuild** with **MSBuild**, however again to no avail. When we tried to install the application using the generated msixbundle, it says that this application is **not trustworthy**.
We also tried explicitly using the VSBuild Task Version that was used in the last successful build, which did not solve the problem either.

That being said, we tried with the same certificate and local msbuild on the local machine and everything worked properly and as expected.

**PS:** We have always been using the same approach to sign our UWP Application with Azure Pipelines and it used to work fine. The last successful build we had was on 25. of February and the first one failing happened on 31. of March. There were no significant code changes in between those builds and we even tried reverting them.

**Question, Bug, or Feature?**:  
Bug  

**Virtual environments affected**  
- [ ] Ubuntu 16.04
- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11.0
- [ ] Windows Server 2016 R2
- [x] Windows Server 2019

**Image version**
20210316.1

**Expected behavior**  
The Application is signed with the provided pfx file.

**Actual behavior**  
The Application seems to be self-signed with the information from the appxmanifest file.

**Comment by Darleev** [comment]
Hello @luEbe,
Thank you for contacting us.
We will check the issue.

**Comment by dsame** [comment]
@luEbe we are not able to reproduce the problem from the ground. Can you please provide a reference to the succeeded and failed builds? Also it would be very useful if there were minimal repro.

**Comment by luEbe** [comment]
@dsame What kind of reference would you like to have? The logs or a URL to the exact build or something else?
Also, we do not have the "succeeded" build anymore, as it was deleted due to our retention policy. Furthermore, the build was always successful, it just did not sign the application as expected.

For the minimal repro, I am not sure what to provide, as we can not say for sure what is causing the problems.

**Comment by dsame** [comment]
@luEbe if your repository is open source the link to it is enough. In the other case please attach the logs.

**Comment by luEbe** [comment]
[logs_2118.zip](https://github.com/actions/virtual-environments/files/6335782/logs_2118.zip)

That is what I can offer. These are the logs of one of the "failing" builds. As I already mentioned, I unfortunately cannot supply you with those of the successful one, as it was deleted.

**Comment by dsame** [comment]
@luEbe 
Ok, i confirm there's no problem with the build steps and they look good

The UWP platform had been upgraded twice between the failed and the succeeded builds so we can not identify the exact source of the problem having the logs only

1. Can you please confirm the UWP components versions used for the local build are the same the ones used in the pipeline?
2. Can you please create the repro using some UWP sample from this repo https://github.com/microsoft/Windows-universal-samples:

1. create a fake cert and add it to the repo
2. create minimal pipeline that mimics your current pipeline
3. add a step to demonstrate the produced binary is not signed properly 

Thanks

**Comment by luEbe** [comment]
@dsame 

Regarding Point 1: What exactly do you mean by UWP components versions? We are using the exact same code base so there should not be any differences, am I right?

We just compared our project file to other UWP Projects and realized, that we are using the **DotnetNativeToolchain**. This has been like that from the beginning of the project and it never caused a problem.
Out of curiosity we tried setting `<UseDotNetNativeToolchain>false</UseDotNetNativeToolchain>` . It now works and it is a validly signed application again (while still using the exact same pfx file).

Also creating a repro is not really an option, as we do not know what exactly is causing the issue and a fresh UWP application would likely work. Furthermore we can not create a fake cert file that reflects the structure of the actual one, as it is signed by our own root authority.

**PS**: We would actually prefer to set `<UseDotNetNativeToolchain>` to true if possible.

**Comment by dsame** [comment]
Please compare the versions of the following components used in the current Win 2019 image with the ones installed in your environment

```
Microsoft.VisualStudio.Component.UWP.VC.ARM64 | 16.3.29207.166
Microsoft.VisualStudio.ComponentGroup.UWP.NetCoreAndStandard | 16.3.29102.218
Microsoft.VisualStudio.ComponentGroup.UWP.Support | 16.9.31004.209
Microsoft.VisualStudio.ComponentGroup.UWP.VC | 16.9.31004.209
Microsoft.VisualStudio.ComponentGroup.UWP.Xamarin
```
(see https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md)

**Comment by luEbe** [comment]
@dsame 
I updated the versions to match the ones installed in the Win 2019 environment.

After that I tried publishing the app via Visual Studio and it worked as expected. Afterwards I executed the exact same command we are using in our Pipeline on my local command line with the **same result as in the pipeline**! The App was untrusted.

I investigated into the log files of both the MSBuild executed by Visual Studio and the one execute from the command line.

There was one significant difference. The one from VS was skipping a Task called "GenerateDisposableSigningCertificate" because it seems that the MSBuild Argument `/p:BuildAppxUploadPackageForUap` is set to `false` by default.

We are setting the MSBuild Argument `/p:UapAppxPackageBuildMode=StoreUpload`. This causes a disposable SigningCertificate to be used, instead of the given one in the other MSBuild Arguments.

This is what we are now using and it is working again:

```
task: VSBuild@1
  inputs:
    platform: 'x64'
    solution: '<insert-solution-file-name>'
    configuration: '$Release'
    msbuildArgs: '/p:AppxBundlePlatforms="x64"
                  /p:AppxPackageDir="<insert-package-dir>"
                  /p:AppxBundle=Always
                  /p:AppxPackageSigningEnabled=true
                  /p:PackageCertificateThumbprint="<insert-thumbprint>" 
                  /p:PackageCertificateKeyFile="<insert-certfile-path>"
                  /p:PackageCertificatePassword="<insert-password>"'
```

I am not sure if this is wanted or unwanted behavior of MSBuild and I will follow up on that with them.

Thank you very much for your help! I consider this issue closed now.
